content/en/blog/_posts/2025-04-21-kubernetes-network-evolution/index.md

```diff
@@ -0,0 +1,71 @@
+
+## The Genesis: Choosing CNI Over CNM
+
+In the early days of Kubernetes, the question of how to manage container networks was a significant one. Two primary models emerged: the Container Network Interface (CNI) and Docker's Container Network Model (CNM). Ultimately, the Kubernetes community opted for CNI, a decision rooted in fundamental architectural and practical considerations, as explained in Tim Hockin’s blog [Why Kubernetes Doesn't use libnetwork](https://blog.kubernetes.io/2016/01/why-kubernetes-doesnt-use-libnetwork/).
+
+The core principle behind CNI was to provide a flexible and focused interface for network plugins. The emphasis was on a standardized mechanism that allowed various network providers to integrate seamlessly with Kubernetes by implementing a simple set of executables. Early discussions in 2015 within the [`kubernetes-sig-network` mailing list](https://groups.google.com/g/kubernetes-sig-network/c/5MWRPxsURUw) reveal a strong desire for stable base plugins that could be extended to meet diverse networking needs. The pluggable design of CNI allowed for the integration of various network plugins, supporting different network topologies and requirements within Kubernetes.
+
+## The Road to Declarative Networking: Key Milestones
+
+With CNI established as the foundation, the Kubernetes community began to work on more advanced networking requirements, particularly the need for a more declarative networking support. This journey was marked by several key initiatives and discussions:
+
+### The Network Plumbing Working Group (NPWG)
+
+The Network Plumbing Working Group (NPWG) was formed at Kubecon NA '17 in Austin to initially work towards implementing a common standard addressing how one might attach multiple networks to pods in Kubernetes. The work of the NPWG has culminated in a [standard Multi-Network Custom Resource Definition (CRD)](https://github.com/k8snetworkplumbingwg/multi-net-spec). The [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) is the reference implementation of this multi-network definition, demonstrating how pods can have multiple network interfaces.
+
+### KEP-3698 (Multi-Network): Reviving the Discussion
+
+The first significant step towards directly addressing multi-network requirements within the core Kubernetes project was the introduction of [KEP-3698 (Multi-Network)](https://github.com/kubernetes/enhancements/pull/3700). This Kubernetes Enhancement Proposal aimed to define user stories and requirements for multi-network functionality in Kubernetes. However, early feedback indicated that the KEP was perceived as too abstract and broad, attempting to encompass a wide range of use cases, from simply attaching multiple interfaces to more complex network virtualization scenarios.
```

Comment:
This immediately raises the question of why Multus/NAD couldn't be brought in-tree.

Reply:
one of the problem of the KEP is that tried to do a lot of things, it had multus as one use case but there were a disparate number of other things that wanted to support to ... included to make kubernetes like a virtual router

Reply:
See, this is why you need more line breaks; I was commenting on the first sentence, not the second. Why did people write a new Multi-Network KEP rather than trying to bring Multus/NAD in-tree? (I know the answer, but the reader won't.)

Reply:
I changed some config on my editor and broke that, I fixed the line breaks no ... give me some suggestion for that part